apps/yggdrasil/lib/yggdrasil/user.ex

```diff
@@ -29,14 +37,120 @@ defmodule Yggdrasil.User do
+
+    permissions = query
+    |> Repo.all()
+    |> Enum.flat_map(fn ur -> ur.role.role_permissions end)
+    |> Enum.group_by(fn rp -> rp.resource end)
+    |> Enum.map(fn {r, rps} -> {String.to_atom(r), Enum.map(rps, fn rp -> String.to_atom(rp.permission) end)} end)
+    |> Enum.into(%{})
+
+    %{ user | permissions: permissions}
+  end
+
+  @doc ~S"""
+  Checks the list of resource/permissions pair given against what the user
+  has been granted and returns true if and only if all permissions supplied
+  have been granted for the user.
+
+  ## roles
+  A user can have one or more roles with each role having a set of resources
+  and a set of resources and those resources having a set of permissions. Roles
+  are not checked directly and only serve as a grouping mechanism for resources.
+
+  Since the user can have multiple roles, the permission set checked for a given
+  resource is the unique set across all the roles for that resource.
+
+  So if the user has `Role1` and `Role2` defined as follows:
+
+  * Role1
+    * foo
+      * :read
+      * :write
+  * Role2
+    * foo
+      * :read
+      * :all
+
+  The set of permissions checked for foo will be `[:read, :write, :all]`
+
+  ## Examples
+      # checking a single resource
+      User.is_granted? user, foo: [:read]
+
+      # checking multiple resources is allowed as well
+      User.is_granted? user, foo: [:read], bar: [:read, :write]
   """
-  def hash_password(changeset = %{:valid? => false}) do
+  def is_granted!(user, res_perms) do
+    case is_granted?(user, res_perms) do
+      {:ok, result}  -> result
+      {:error, msgs} -> raise RuntimeError, inspect(msgs)
+    end
+  end
+
+  def is_granted?(_user, []),  do: build_error ["resource keyword list is empty"]
+  def is_granted?(_user, nil), do: build_error ["resource keyword list is nil"]
+  def is_granted?(user, res_perms) do
+    res_perms
+    |> Enum.map(fn {res, perms} -> {res, user.permissions[res], perms} end)
+    |> Enum.map(&has_permissions/1)
+    |> Enum.reduce({:ok, true}, fn
+      {:ok, _},  acc={:ok, false}   -> acc
+      {:ok, res},    {:ok, true}    -> {:ok, res}
+      {:ok, _},  acc={:error, _}    -> acc
+      {:error, msg}, {:ok, _}       -> {:error, [msg]}
+      {:error, msg}, {:error, msgs} -> {:error, [msg|msgs]}
+    end)
+  end
+
+  defp has_permissions({_res, nil, _requried}), do: {:ok, false}
+  defp has_permissions({res, _given, nil}),     do: build_error "required permission list is nil for #{res}"
+  defp has_permissions({res, [], _required}),   do: build_error "given permission list is empty for #{res}"
+  defp has_permissions({res, _given, []}),      do: build_error "required permission list is empty for #{res}"
+  defp has_permissions({_res, given, required}) do
+    given = MapSet.new given
+    required = MapSet.new required
+
+    {:ok, MapSet.subset?(required, given)}
+  end
+
+  defp build_error(msg) do
+    Logger.error msg
+    {:error, msg}
+  end
```

Comment:
I still feel like this should just be is_granted?/2 function that raises an error

I don't think there is anything the caller could do or any recovering action. need to dig into ectos repo but I think for example even if you are using like Repo.all or the others that if there is db connection issue that it raise an exception instead of returning an error tuple. Think I witnessed it yesterday actually. 

I think error tuple are perfect for things that could fail when called normally, like trying to read a file or insert a record where handling the error for reading a file you may retry x number of times and with insert maybe it fails because of a unique constraint which the user will be able to correct by choosing a unique name. 

With is_granted?/2 you don't have that pattern really. If it raises an error it was called wrong/configured wrong and truly is a bad state that should halt and I think that is ok and idiomatic. Not everything uses error tuples. 

Don't know happy to leave as is for now, but I think I was on the right track with no error tuples and simply returned true or false and raised errors in those bad states.